contributors/design-proposals/storage/local-storage-pv.md

```diff
@@ -761,3 +761,124 @@ The central manager will poll each worker for their discovered volumes and creat
+which depends on the newly introduced changes in [Volume Topology-aware Scheduling](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md),
+that is:
+
+1. When feature `DynamicProvisioningScheduling` enabled, scheduler would verify a pod's volume requirement,
+and set `annScheduledNode` on PVCs that need provisioning.
+2. New fields would be added in `StorageClass` to allow storage providers to report
+capacity limits per topology domain.
+
+The local volume provisioner need to be updated in several aspects to support the scenario.
+
+##### Backend storage management
+
+Mechanisms to manage local storage sources are varied. They could be LVM, Ceph, or something like that.
+In addition to current `populator`, `deletor` and `discoverer`, a new component named
+`storageManager` will be introduced in the plugin, to manage the backend storage.
+
+`storageManager` is a general interface, each kind of storage should have its own implementation.
+During start-up, a new environment variable `STORAGE_BACKEND` is set in the provisioner DaemonSet,
+to determine which StorageManager instance to start.
+(TBD: Should we support the scenario of multi backends on a node?)
+
+A StorageManager instance is responsible for:
+
+1. Monit changes in storage input, and handle corresponding initialization.
+2. Update capacity in StorageClass when storage source changed.
+3. Provide the ability to create volumes, detailed in the volume creation section.
+
+Take LVM as an example, when a new source disk "/dev/new" is added to its StorageClass `exStorage`,
+the StorageManager should:
+
+1. Check if related PV (Physical volume in LVM) exists, if not, create one with command:
+```
+pvcreate /dev/new
+```
+
+2. Check if the Physical volume exists in Volume group exStorageGroup, if not, add it:
+```
+vgextend exStorageGroup /dev/new
+```
+
+3. Update capacity in StorageClass if there is a change in VG size.
+
+A new field `storageSource` will be added into the configuration ConfigMap, to illustrate available
+storage source for provisioning.
+
+```
+kind: ConfigMap
+metadata:
+  name: local-volume-config
+  namespace: kube-system
+data:
+ storageClassMap: |
+    local-fast:
+      hostDir: "/mnt/ssds"
+    local-slow:
+      hostDir: "/mnt/hdds"
+    local_dynamic:
+      storageSource:
+        - "/dev/sdb"
+        - "/dev/sdc"
+```
+For LVM, its items are disk partitions, and for other mechanisms, they could be other things.
+StorageManager should know the scope and what to do with the input.
+
+Major downsides of this approach:
+
+* As provisioners running on all nodes shall a same configuration ConfigMap,
+storage sources of their nodes are required to be configured in the same way.
+Take LVM as an example, for configmap above, disks used for provisioning are required
+to be partitioned as "/dev/sdb" and "/dev/sdc".

```

Comment:
I think this situation could be handled via multiple means.

Cluster operators could leverage multiple dameonset manifests and various selectors. For example, if you had two different types of hardware with a different number of partitions and different partition locations, then you could label the nodes accordingly and deploy two daemonsets where each daemonset config was tailored toward the appropriately labelled nodes.

Alternatively, another earlier process in the node's lifecycle (e.g. nodeprep - https://engineering.salesforce.com/provisioning-kubernetes-local-persistent-volumes-61a82d1d06b0) could perform operations before the lv dynamic provisioner is used, and the node preparation could make nodes fit the common configuration via symlinks or bindmounts.

Reply:
Agree, we should support the use case of different number and path to disks on every node, along with disks being dynamically added and removed. Maybe another discovery directory or local file is needed.

Reply:
Supporting wildcards or regex in the storageSources listed could provide some additional flexibility to support multiple node's and their individual configurations.

Reply:
Really appreciate for you guys suggestions, I will look into that.

Reply:
Really appreciate for you guys suggestions, I will look into that.